state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -3019,81 +3010,70 @@ pub mod pallet {
-			})
-		}
-
 		fn calculate_input_for_desired_output(
 			input_asset: Asset,
 			output_asset: Asset,
 			desired_output_amount: AssetAmount,
 			with_network_fee: bool,
-		) -> Option<AssetAmount> {
-			use frame_support::sp_runtime::helpers_128bit::multiply_by_rational_with_rounding;
+		) -> AssetAmount {
+			// Approximate one sided slippage adjustments for oracle prices
+			const ORACLE_SLIPPAGE: BasisPoints = 40;
+			const ORACLE_SLIPPAGE_STABLE: BasisPoints = 3;
 
 			if desired_output_amount.is_zero() {
-				return Some(Zero::zero())
+				return 0;
 			}
 
+			let network_fee = if with_network_fee {
+				let fee_rate_and_minimum =
+					Pallet::<T>::get_network_fee_for_swap(input_asset, output_asset, false);
+				// Ignoring the minimum network fee because this function is only used for fees and
+				// gas (no minimum).
+				fee_rate_and_minimum.rate
+			} else {
+				Permill::zero()
+			};
+
 			if input_asset == output_asset {
-				return Some(desired_output_amount)
+				// Network fee will still be taken on a same asset swap
+				return desired_output_amount + network_fee * desired_output_amount;
 			}
 
-			let estimation_input = utilities::estimated_20usd_input(input_asset);
-
-			let estimation_output = with_transaction_unchecked(|| {
-				TransactionOutcome::Rollback(if with_network_fee {
-					Self::swap_with_network_fee_for_gas(input_asset, output_asset, estimation_input)
-						.map(|swap| swap.output)
-				} else if output_asset == STABLE_ASSET || input_asset == STABLE_ASSET {
-					T::SwappingApi::swap_single_leg(input_asset, output_asset, estimation_input)
+			// Get the price of both assets using oracles or simulated swap, both with fallback to
+			// hard coded prices
+			let get_price = |asset, side: Side| -> Price {
+				if asset == Asset::Flip || asset == Asset::Dot {
+					Self::estimate_price_using_simulated_swap(input_asset, side)
 				} else {
-					T::SwappingApi::swap_single_leg(input_asset, STABLE_ASSET, estimation_input)
-						.and_then(|intermediary| {
-							T::SwappingApi::swap_single_leg(
-								STABLE_ASSET,
-								output_asset,
-								intermediary,
-							)
-						})
-				})
-			})
-			.ok()?;
-
-			if estimation_output == 0 {
-				None
-			} else {
-				let input_amount_to_convert = multiply_by_rational_with_rounding(
-					desired_output_amount,
-					estimation_input,
-					estimation_output,
-					sp_arithmetic::Rounding::Down,
-				)
-				.defensive_proof(
-					"Unexpected overflow occurred during asset conversion. Please report this to Chainflip Labs."
-				)?;
-				if input_amount_to_convert.is_zero() {
-					None
-				} else {
-					Some(input_amount_to_convert.unique_saturated_into())
+					// Using stale prices here is fine as its just for fees/gas
+					T::PriceFeedApi::get_price(asset).map(|price_data| {
+						// Apply a hard coded slippage in the correct direction
+						let slippage_bps = if asset == STABLE_ASSET {
+							0
+						} else if asset == Asset::Usdc ||
```

Comment:
We don't need to list `Asset::Usdc` here, we already checked it above (as `STABLE_ASSET`).